lib/ecto/changeset.ex

```diff
@@ -1012,7 +1012,20 @@ defmodule Ecto.Changeset do
           "please use put_#{tag}/4 instead"
   end
 
-  defp put_change(data, changes, errors, valid?, key, value, _type) do
+  defp put_change(data, changes, errors, valid?, key, value, nil) do
+    put_change(data, changes, errors, valid?, key, value)
+  end
+
+  defp put_change(data, changes, errors, valid?, key, value, type) do
+    case Ecto.Type.cast(type, value) do
+      {:ok, value} ->
+        put_change(data, changes, errors, valid?, key, value)
+      :error ->
+        raise Ecto.CastError, "cannot cast #{inspect value} to type #{inspect type}"
```

Comment:
if accepted, we could also use Ecto.Type.Cast proposed in https://github.com/elixir-ecto/ecto/pull/2031/files#diff-20335f2fd54513fa94ed285d49d98147R130